moonraker/components/prometheus.py

```diff
@@ -0,0 +1,157 @@
+# Prometheus client implementation for Moonraker
+#
+# Copyright (C) 2024 Kamil Domański <kamil@domanski.co>
+#
+# This file may be distributed under the terms of the GNU GPLv3 license.
+
+from __future__ import annotations
+from asyncio import gather
+import logging
+from prometheus_client import (
+    exposition, registry,
+    Info, Gauge
+)
+from prometheus_client.metrics import MetricWrapperBase
+
+from ..common import (
+    KlippyState,
+    RequestType,
+    TransportType,
+    WebRequest
+)
+
+from .klippy_apis import KlippyAPI
+
+from typing import (
+    TYPE_CHECKING,
+    Dict,
+    Any
+)
+
+if TYPE_CHECKING:
+    from ..confighelper import ConfigHelper
+
+class PrometheusExporter:
+    def __init__(self, config: ConfigHelper) -> None:
+        self.server = config.get_server()
+        app_args = self.server.get_app_args()
+
+        # Not saved in the object, because it should not be changed or cleared.
+        i = Info('moonraker_instance', '')
+        i.info({
+            'version': app_args['software_version'],
+            'instance_uuid': app_args['instance_uuid'],
+            'python_version': app_args['python_version'],
+        })
+
+        # metrics
+        self.m_temp = Gauge(
+            'temp', 'Current temperature of a heater or sensor', ['sensor'])
+        self.m_target_temp = Gauge(
+            'target_temp', 'Target temperature of a heater or fan', ['sensor'])
+        self.m_heater_power = Gauge(
+            'heater_power', 'Current power setting of a heater', ['heater'])
+
+        self.server.register_endpoint(
+            "/server/prometheus/metrics", RequestType.GET,
+            self._handle_metrics_endpoint, transports=TransportType.HTTP,
+            wrap_result=False, content_type=exposition.CONTENT_TYPE_LATEST
+        )
+        self.server.register_event_handler(
+            "server:klippy_started", self._handle_klippy_started
+        )
+
+        # clear metrics to stop providing metrics when their value is simply unknown
+        self.server.register_event_handler(
+            "server:klippy_shutdown", self._clear_metrics
+        )
+        self.server.register_event_handler(
+            "server:klippy_disconnect", self._clear_metrics
+        )
+        self.server.register_event_handler(
+            "server:klippy_disconnected", self._clear_metrics
+        )
```

Comment:
Moonraker doesn't send a `klippy_disconnected` event.  It does register a `notify_klippy_disconnected` websocket notification that piggybacks off of the `server:klippy_disconnect` event.